[CI:DOCS] s3fs volume mount docs (#21620)
containers/podman: docs/source/markdown/podman-volume-create.1.md

All volume assigned project IDs larger project IDs starting with 200000.
This prevents xfs_quota management conflicts with containers/storage.

## MOUNT EXAMPLES

> podman volume create allows the `type`, `device`, and `o` options to be passed to `mount(8)` when using the `local` driver.

Reviewer: ```suggestion
The `podman volume create` command allows the `type`, `device`, and `o` options to be passed to `mount(8)` when using the `local` driver.
```

Author: I updated the way it was written under QUOTAS , and under my changes.

Reviewer: @WesselAtWork, good point about "The ... command".  It is a bit redundant, and when I took a second look at the whole page, I realized we'd either need to drop it here or add it in a number of other places to be consistent.  Once added, it would be TERRIBLY redundant.  Great pushback; please do drop my "The ... command" and it looks like you have. :+1:

## [s3fs-fuse](https://github.com/s3fs-fuse/s3fs-fuse)

[s3fs-fuse](https://github.com/s3fs-fuse/s3fs-fuse) or just `s3fs`, is a [fuse](https://github.com/libfuse/libfuse) filesystem that allows s3 prefixes to be mounted as filesystem mounts.

**Installing:**
```shell
$ doas dnf install s3fs-fuse
```

**Simple usage:**
```shell
$ s3fs --help
$ s3fs -o use_xattr,endpoint=aq-central-1 bucket:/prefix /mnt
```

**Equivalent through `mount(8)`**
```shell
$ mount -t fuse.s3fs -o use_xattr,endpoint=aq-central-1 bucket:/prefix /mnt
```

**Equivalent through `podman volume create`**
```shell
$ podman volume create s3fs-fuse-volume -o type=fuse.s3fs -o device=bucket:/prefix -o o=use_xattr,endpoint=aq-central-1
```

**The volume can then be mounted in a container with**
```shell
$ podman run -v s3fs-fuse-volume:/s3:z --rm -it fedora:latest
```

Please see the available [options](https://github.com/s3fs-fuse/s3fs-fuse/wiki/Fuse-Over-Amazon#options) on their wiki.

### Using with other container users

The above example works because the volume is mounted as the host user and inside the container `root` is mapped to the user in the host.

> If the mount is accessed by a different user inside the container, a "Permission denied" error will be raised. Even though that user should be mapped to one of the caller's `subuid`.

Reviewer 2: Sentence fragment. After thinking about it, I recommend just removing the entire "Even though ... subuid" fragment, it doesn't add anything. Just the first sentence ("If the mount ... will be raised") is sufficient.

Author: I'd argue that it gets the reader to think about how a `subid` interacts with the host.
Hindsight, it's obvious, but when I was figuring this out it wasn't so. Working on this gave me a "concrete" understanding of what user is actually "inside" / "outside" the container.
The hope would be to help someone else "get-it"

But I agree this sentence isn't great, and the value of the above is questionable at best.
Removed.

> If the mount is accessed by a different user inside the container, a "Permission denied" error will be raised. Even though that user should be mapped to one of the caller's `subuid`'s.

Reviewer 2: No apostrophe, just backtick-subuid-backtick-s

```shell
$ podman run --user bin:bin -v s3fs-fuse-volume:/s3:z,U --rm -it fedora:latest
$ ls /s3
# ls: /s3: Permission denied
```

> In FUSE-land, mounts are protected for the user who mounted them; Spesify the `allow_other` mount option if other users need access too.

Reviewer: ```suggestion
In FUSE-land, mounts are protected for the user who mounted them; specify the `allow_other` mount option if other users need access.
```

Reviewer: I like your rewording, it's a lot less awkward.  One typo, and a couple of grammatical nits.

> In FUSE-land, mounts are protected to the user who mounted it; If another users needs to be able to access this mount, specify the `allow_other` mount option.

Reviewer: ```suggestion
In FUSE-land, mounts are protected for the user who mounted them. If another user needs to be able to access this mount, specify the `allow_other` mount option.
```

Author: That sentence was actually very awkward. I reworded it, please check the reword.

> Note: This will remove the normal fuse [security measures](https://github.com/libfuse/libfuse/wiki/FAQ#why-dont-other-users-have-access-to-the-mounted-filesystem) on the mount point, after which, the normal filesystem permissions will have to protect it

```shell
$ podman volume create s3fs-fuse-other-volume -o type=fuse.s3fs -o device=bucket:/prefix -o o=allow_other,use_xattr,endpoint=aq-central-1
$ podman run --user bin:bin -v s3fs-fuse-volume:/s3:z,U --rm -it fedora:latest
$ ls /s3
```

### The Prefix must exist

`s3fs` will fail to mount if the prefix does not exist in the bucket.

> You can create a "directory" by putting an empty object at the desired `prefix/` key

Reviewer 2: One of our maintainers will pounce on this because we can't use the word "you". Please rewrite. Also, no quotes needed around `directory`

All volume assigned project IDs larger project IDs starting with 200000.
This prevents xfs_quota management conflicts with containers/storage.

## MOUNT EXAMPLES

> podman volume create allows you to pass the `type`, `device`, and `o` options to `mount(8)` when using the `local` driver.  

Reviewer: ```suggestion
podman volume create allows the `type`, `device`, and `o` options to be passed to `mount(8)` when using the `local` driver.  
```

> Here are some examples of what you can do with that

Reviewer 2: Given the prominent heading `MOUNT EXAMPLES`, I'm not sure why/if these two lines are needed.

Reviewer: I concur, I'd drop line 136.

Author: Wanted to separate it from the other examples, like previously mentioned, this is a library that is not controlled by this project, but because it can interface with `mount` it can integrate with other libraries.  
Wanted to also have a new heading for future expansion.  

Some ideas:
Maybe someone figures something out with that shows off a capability, like: mounting a [iSCSI](https://github.com/sahlberg/libiscsi) target, an [iPFS](https://github.com/ipfs/kubo/blob/master/docs/fuse.md) target or mounting a weird device like USB for use with USB/IP, an example with using [bindfs](https://bindfs.org/) that allows you to do something cool with mirrors, maybe mounting [archives](https://github.com/google/fuse-archive)  

---

If it is still out of place, where would you like me to put it?  
Under the existing examples?

Author: Ah wait, I re-read your comment, I missunderstood what you meant. I agree, it doesn't read very nice.

## [s3fs-fuse](https://github.com/s3fs-fuse/s3fs-fuse)

> [s3fs-fuse](https://github.com/s3fs-fuse/s3fs-fuse) or just `s3fs`, is a fuse filesystem that allows you to mount s3 prefixes as filesystem mounts.

Reviewer: ```suggestion
[s3fs-fuse](https://github.com/s3fs-fuse/s3fs-fuse), or just `s3fs`, is a fuse filesystem that allows s3 prefixes to be mounted as filesystem mounts.
```

**Installing:**
```shell

> dnf install s3fs-fuse

Reviewer 2: Always include a prompt. Always. In this case, prompt should be `#` to indicate that the command must be run as root. Alternatively, use `$ sudo ...`.

# should also install fuse3 as a dependency
```

**Simple usage:**
```shell
s3fs --help
s3fs -o use_xattr,endpoint=aq-central-1 bucket:/prefix /mnt
```

> **Eqivelent through `mount(8)`**

Reviewer 2: Typo, "Equivalent"

```shell
mount -t fuse.s3fs -o use_xattr,endpoint=aq-central-1 bucket:/prefix /mnt
```

> **Eqivelent through `podman volume create`**

Reviewer: ```suggestion
**Equivalent through `podman volume create`**
```

```shell
podman volume create s3fs-fuse-volume -o type=fuse.s3fs -o device=bucket:/prefix -o o=use_xattr,endpoint=aq-central-1
```

**You can then mount it in a container with**
```shell
podman run -v s3fs-fuse-volume:/s3:z --rm -it fedora:latest
```

Please see the available [options](https://github.com/s3fs-fuse/s3fs-fuse/wiki/Fuse-Over-Amazon#options) on their wiki.

### Using with other container users

The above example works because the volume is mounted as the host user and inside the container `root` is mapped to the user in the host. 

> If you tried to access the mount with a different user inside the container you will be hit with a "Permission denied"

Reviewer: ```suggestion
If the mount is accessed by a different user inside the container, a "Permission denied" error will be raised.
```

```shell
podman run --user bin:bin -v s3fs-fuse-volume:/s3:z,U --rm -it fedora:latest
ls /s3
# ls: /s3: Permission denied
```

> Even though that user should be maped to one of your `subuid`'s  

Reviewer: ```suggestion
Even though that user is mapped to one of the `subuid`'s  
```

> In FUSE-land mounts are protected to the user who mounted it, if you want other users to be able to access these mounts, you need to specify the `allow_other` mount option.

Reviewer 2: In FUSE-land **COMMA** mounts ... mounted it **PERIOD**.

And, our guidelines prevent use of the word "you". Can you please phrase this differently?

> In FUSE-land mounts are protected to the user who mounted it, if you want other users to be able to access these mounts, you need to specify the `allow_other` mount option.

Reviewer: ```suggestion
In FUSE-land, mounts are protected by the user who mounted them; if another user needs to be able to access these mounts, specify the `allow_other` mount option.
```

> Note: This will remove the normal fuse security measures on the mount point, and you will have to rely on fs-permisions to protect it

```shell
podman volume create s3fs-fuse-other-volume -o type=fuse.s3fs -o device=bucket:/prefix -o o=allow_other,use_xattr,endpoint=aq-central-1
podman run --user bin:bin -v s3fs-fuse-volume:/s3:z,U --rm -it fedora:latest
ls /s3
```

### The Prefix must exist

`s3fs` will fail to mount if the prefix does not exist in the bucket.   

> You can very simply create it by putting an empty object at the `prefix/` key  

Reviewer: ditto you
```suggestion
Simply create it by putting an empty object at the `prefix/` key  
```

> this is how aws denotes "directories"
```shell
aws s3api put-object --bucket bucket --key prefix/
```

> If performance is your priority, please check out the more performant [goofys](https://github.com/kahing/goofys)  

Reviewer: ```suggestion
If performance is the priority, please check out the more performant [goofys](https://github.com/kahing/goofys)  
```
